shim.c

```diff
@@ -1895,25 +1896,20 @@ efi_main (EFI_HANDLE passed_image_handle, EFI_SYSTEM_TABLE *passed_systab)
 	 */
 	debug_hook();
 
-	INIT_LIST_HEAD(&sbat_var);
-	efi_status = parse_sbat_var(&sbat_var);
-	/*
-	 * Until a SBAT variable is installed into the systems, it is expected that
-	 * attempting to parse the variable will fail with an EFI_NOT_FOUND error.
-	 *
-	 * Do not consider that error fatal for now.
-	 */
-	if (EFI_ERROR(efi_status) && efi_status != EFI_NOT_FOUND) {
-		perror(L"Parsing SBAT variable failed: %r\n",
-		       efi_status);
-		msg = IMPORT_SBAT;
-		goto die;
-	}
-
-	if (secure_mode ()) {
+	if (secure_mode()) {
 		char *sbat_start = (char *)&_sbat;
 		char *sbat_end = (char *)&_esbat;
 
+		INIT_LIST_HEAD(&sbat_var);
+		efi_status = parse_sbat_var(&sbat_var);
+		if (EFI_ERROR(efi_status)) {
+			perror(L"Parsing SBAT variable failed: %r\n",
+					efi_status);
+			msg = IMPORT_SBAT;
+			efi_status = set_sbat_uefi_variable();
```

Comment:
The return status of set_sbat_uefi_variable() wasn't checked and it caused the strange failure during the first boot:

shim.c:1891:efi_main() vendor_deauthorized:0x7DDDB4F9 vendor_deauthorized_size:10684
Parsing SBAT variable failed: Not Found
sbat.c:309:set_sbat_uefi_variable() SBAT read failed Not Found
sbat.c:326:set_sbat_uefi_variable() SBAT read failed Not Found
sbat.c:339:set_sbat_uefi_variable() SBAT variable delete failed Not Found
sbat.c:345:set_sbat_uefi_variable() SBAT variable clearing failed Not Found
sbat.c:372:set_sbat_uefi_variable() SBAT variable initialization succeeded
Something has gone seriously wrong: import of SBAT data failed: Success

Reply:
Thank you! It's not just the missing check, first boot in secure mode isn't right.

This needs to be:
  - try to import / parse sbat
  - try to set variable
  - try to import again

Reply:
@martinezjavier I think I fixed this last night, but can you take another quick look at the ordering?

Reply:
I see why I didn't think I had this quite right last night, adding a check to skip the setting if we succeeded previously.

Reply:
Actually no, this is fine as is now (after fixing the issue pointed out by @lcp). set_sbat_uefi_variable() will run again and see that no work is to be done, or try one more time if the variable magically vanished after previously reading back. (Though that's a super broken system.)